pkg/api/custom_api.go

```diff
@@ -0,0 +1,298 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gorilla/mux"
+	"github.com/grafana/regexp"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/munnerz/goautoneg"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/util/annotations"
+	"github.com/prometheus/prometheus/util/httputil"
+	v1 "github.com/prometheus/prometheus/web/api/v1"
+)
+
+type CustomAPI struct {
+	queryable     storage.SampleAndChunkQueryable
+	queryEngine   promql.QueryEngine
+	now           func() time.Time
+	statsRenderer v1.StatsRenderer
+	logger        log.Logger
+	codecs        []v1.Codec
+	CORSOrigin    *regexp.Regexp
+}
+
+func NewCustomAPI(
+	qe promql.QueryEngine,
+	q storage.SampleAndChunkQueryable,
+	statsRenderer v1.StatsRenderer,
+	logger log.Logger,
+	codecs []v1.Codec,
+	CORSOrigin *regexp.Regexp,
+) *CustomAPI {
+	return &CustomAPI{
+		queryable:     q,
+		queryEngine:   qe,
+		now:           time.Now,
+		statsRenderer: statsRenderer,
+		logger:        logger,
+		codecs:        codecs,
+		CORSOrigin:    CORSOrigin,
+	}
+}
+
+// Register custom api path with handler
+// Usage: router.Path(path.Join(prefix, ${path})).Methods(${method}).Handler(${handler})
+func (c *CustomAPI) Register(router *mux.Router, prefix string) {
+	router.Path(path.Join(prefix, "/api/v1/query")).Methods("GET", "POST").Handler(c.wrap(c.instantHandler))
+	router.Path(path.Join(prefix, "/api/v1/query_range")).Methods("GET", "POST").Handler(c.wrap(c.rangeQueryHandler))
+}
+
+// Custom handler for Query range API
+func (c *CustomAPI) rangeQueryHandler(r *http.Request) (result apiFuncResult) {
```

Comment:
I would suggest we move query handlers implementation to its own package in `pkg/querier`.

Reply:
@SungJin1212 Can we change the file path as Harry suggested?